SteamKit2/SteamKit2/Steam/CMClient.cs

```diff
@@ -212,26 +212,32 @@ public void Connect( ServerRecord? cmServer = null )
                         return;
                     }
 
-                    connection = CreateConnection( record.ProtocolTypes & Configuration.ProtocolTypes );
-                    connection.NetMsgReceived += NetMsgReceived;
-                    connection.Connected += Connected;
-                    connection.Disconnected += Disconnected;
-                    connection.Connect( record.EndPoint, ( int )ConnectionTimeout.TotalMilliseconds );
+                    var newConnection = CreateConnection( record.ProtocolTypes & Configuration.ProtocolTypes );
+
+                    var connectionRelease = Interlocked.Exchange( ref connection, newConnection );
+                    Debug.Assert( connectionRelease == null );
+
+                    newConnection.NetMsgReceived += NetMsgReceived;
+                    newConnection.Connected += Connected;
+                    newConnection.Disconnected += Disconnected;
+                    newConnection.Connect( record.EndPoint, ( int )ConnectionTimeout.TotalMilliseconds );
                 }, TaskContinuationOptions.ExecuteSynchronously ).ContinueWith( t =>
-              {
+                {
                     if ( t.IsFaulted )
                     {
-                      LogDebug( nameof( CMClient ), "Unhandled exception when attempting to connect to Steam: {0}", t.Exception );
+                        LogDebug( nameof( CMClient ), "Unhandled exception when attempting to connect to Steam: {0}", t.Exception );
                         OnClientDisconnected( userInitiated: false );
                     }
+
+                    connectionSetupTask = null;
```

Comment:
I think that's the right place to put it? We don't want this task reference to linger after it's done.

Reply:
This can't be done as Disconnect may be trying to access it.

Reply:
This would be nulled after the task is already finished its work, does it matter if disconnect accesses it?